File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
from collections import deque

def solution(msg):
    answer = []
    dictionary = {}
    msg = deque(list(msg))
    
    index = 0
    for i in range(ord('A'), ord('Z') + 1):
        index += 1
        dictionary[chr(i)] = index
    
    while msg:
        w = msg[0]
        
        i = 1
        while i < len(msg) and w + msg[i] in dictionary:
            w += msg[i]
            i += 1
        
        answer.append(dictionary[w])
        
        for _ in range(len(w)):
            msg.popleft()
        
        if len(msg) > 0:
            index += 1
            dictionary[w + msg[0]] = index

    return answer

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;

class Solution {
    public int[] solution(String msg) {
        var answerArray = new ArrayList<Integer>();
        var dictionary = new HashMap<String, Integer>();
        
        int index = 1;
        for (int i = (int)'A'; i <= (int)'Z'; i++) {
            dictionary.put(Character.toString((char)i), index++);
        }
        
        int pointer = 0;
        String token = "";
        while (pointer < msg.length()) {
            String cur = Character.toString(msg.charAt(pointer));
            
            if (dictionary.containsKey(token + cur) && pointer == msg.length() - 1) {
                answerArray.add(dictionary.get(token + cur));
                break;
            }
            
            // 뒷글자까지 이어서 사전에 있으면 계속 진행
            // 사전에 없으면 그 전까지 토큰 번호 answer에 추가, 현재 토큰은 dict에 추가
            if (dictionary.containsKey(token + cur)) {
                token += cur;
                pointer += 1;
            } else {
                answerArray.add(dictionary.get(token));
                dictionary.put(token + cur, index++);
                token = "";
            }
        }
        
        return answerArray.stream().mapToInt(Integer::intValue).toArray();
    }
}